sirepo/package_data/static/js/sirepo-components.js

```diff
@@ -884,6 +902,31 @@ SIREPO.app.directive('msieFontDisabledDetector', function(errorService, $interva
     };
 });
 
+SIREPO.app.directive('safePath', function() {
+    return {
+        restrict: 'A',
+        require: 'ngModel',
+        link: function(scope, element, attrs, ngModel, $sce) {
+
+            scope.showWarning = false;
+            scope.warningText = scope.info[0] + ' must not include any of the following: ' +
+                SIREPO.SAFE_PATH_REGEXP_EXCLUDE_CHARS.join(' ').replace(/\\\\/, '\\');
+
+            ngModel.$validators.safe = function(value) {
+                var hasInvalidChars = SIREPO.SAFE_PATH_REGEXP_EXCLUDE.test(value);
+                scope.showWarning = hasInvalidChars;
+                return !hasInvalidChars;
+            }
+
+            scope.$on('cancelChanges', function(e, data) {  // broadcast by root scope
+                scope.showWarning = false;
+                ngModel.$setViewValue('');
+                ngModel.$render();  // appears that invalid text will not be cleared unless we do this
```

Comment:
Put this comment above the line as it is important. I would consider encapsulating this, e.g.

```
function clearViewValue(model) {
    model.$setViewValue('');
    model.$render();
}
```

Seems like this would be a general problem we have when we cancel, even resetting all fields.